### ISSUE TITLE
SymbolicPropogator incorrectly "emulates" segment op

### ISSUE THREAD

**Issue by astrelsky** [NationalSecurityAgency/ghidra#4252]
**Describe the bug**
The SymbolicPropogator is still using the hard coded "emulation" for the segment op. This may be wrong, and is wrong, for some processors. It should be using the segment op defined by the processor by using the pcode injection for it.

Arguably all user defined dynamic pcode injection should be injected. 

**Environment (please complete the following information):**
 - Ghidra Version: 10.2 DEV

**Comment by emteere** [comment]
We have been working on fixing the segment emulation issue, the change just hasn't made it into patch yet.
The change will emulate all pcode injection ops including the segment op.
The main holdup up is making sure performance doesn't take a hit, as it has to check every userop.

If you would like, I can post the patch and you can try it out.

**Comment by astrelsky** [comment]
> We have been working on fixing the segment emulation issue, the change just hasn't made it into patch yet.
> The change will emulate all pcode injection ops including the segment op.
> The main holdup up is making sure performance doesn't take a hit, as it has to check every userop.
> 
> If you would like, I can post the patch and you can try it out.

Sure. I wouldn't be able to give it a go until the weekend though.

**Comment by emteere** [comment]
Patch and master has an initial change.
There is another, tracking constants in/out of the stack that has segmented accesses, that didn't make it.
This is prototype, so YMMV

```
diff --git a/Ghidra/Features/Base/src/main/java/ghidra/program/util/SymbolicPropogator.java b/Ghidra/Features/Base/src/main/java/ghidra/program/util/SymbolicPropogator.java
index b3b733a..ca270b4 100644
--- a/Ghidra/Features/Base/src/main/java/ghidra/program/util/SymbolicPropogator.java
+++ b/Ghidra/Features/Base/src/main/java/ghidra/program/util/SymbolicPropogator.java
@@ -979,7 +979,7 @@
 
 					// for callother, could be an interrupt, need to look at it like a call
 					case PcodeOp.CALLOTHER:
-						PcodeOp[] callOtherPcode = doCallOtherPcodeInjection(instruction, in, out);
+						PcodeOp[] callOtherPcode = doCallOtherPcodeInjection(instruction, pcodeOp, in, out);
 						
 						if (callOtherPcode != null) {
 							ops = injectPcode(ops, pcodeIndex, callOtherPcode);
@@ -1651,7 +1651,7 @@
 	 * 
 	 * @throws NotFoundException
 	 */
-	private PcodeOp[] doCallOtherPcodeInjection(Instruction instr, Varnode ins[], Varnode out) throws NotFoundException {
+	private PcodeOp[] doCallOtherPcodeInjection(Instruction instr, PcodeOp op, Varnode ins[], Varnode out) throws NotFoundException {
 		Program prog = instr.getProgram();
 
 		PcodeInjectLibrary snippetLibrary = prog.getCompilerSpec().getPcodeInjectLibrary();
@@ -1665,6 +1665,9 @@
 		for (int i = 1; i < ins.length; i++) {
 			Varnode vval = context.getValue(ins[i], evaluator);
 			if (!vval.isConstant()) {
+				if (i==1) { // first input is an unknown value
+					return injectStackSymbolicCopy(payload, op, out, ins);
+				}
 				return null;
 			}
 			inputs.add(vval);
@@ -1680,7 +1683,49 @@
 		con.output.add(out);
 		return payload.getPcode(prog, con);
 	}
+
+	/**
+	 * Check if this is a stack symbolic segment op, which can be turned into a copy
+	 *   iff the the offset portion of the segment op is symbolic
+	 * @param payload - injection payload
+	 * @param op - callother pcodeop
+	 * @param out - output varnode for callother
+	 * @param ins - inputs to callother
+	 * @return symbolic stack copy, or null if not a stack symbolic segment operation
+	 * @throws NotFoundException
+	 */
+	private PcodeOp[] injectStackSymbolicCopy(InjectPayload payload, PcodeOp op, Varnode out, Varnode[] ins)
+			throws NotFoundException {
+		// if segment op, then check for a symbolic case passthru
+		if (!"segment_pcode".equals(payload.getName())) {
+			return null;
+		}
+		
+		Varnode vval = context.getValue(ins[2], evaluator);
+		// TODO: could pass through any symbolic offset
+		// if (!context.isSymbol(vval)) {
+		//    return null;
+		// }
+		// only pass through stack symbolic offsets
+		if (!context.isStackSymbolicSpace(vval)) {
+			return null;
+		}
+		
+		Varnode nins[] = new Varnode[1];
+		nins[0] = ins[2];
+		PcodeOp ops[] = new PcodeOp[1];
+		ops[0] = new PcodeOp(op.getSeqnum(), PcodeOp.COPY, nins, out);
+		return ops;
+	}
 	
+	/**
+	 * Find the callother pcode injection if it exists
+	 * 
+	 * @param prog program
+	 * @param snippetLibrary pcode injection snippet library
+	 * @param callOtherIndex callother ID
+	 * @return
+	 */
 	private InjectPayload findPcodeInjection(Program prog, PcodeInjectLibrary snippetLibrary, long callOtherIndex) {
 		InjectPayload payload = (InjectPayload) injectPayloadCache.get(callOtherIndex);
 		
```

**Comment by astrelsky** [comment]
Great thank you.  I will give the patch a shot sometime this weekend.

**Comment by astrelsky** [comment]
@emteere at first glance everything seems to be working correctly. Admittedly I'm not super familiar with the instruction set I have here so if something is off it might take me a while to notice.

**Comment by astrelsky** [comment]
```
+				if (i==1) { // first input is an unknown value
+					return injectStackSymbolicCopy(payload, op, out, ins);
+				}
```

I was encountering a problem caused by the above added check. Is that assuming that the segmend op is defined as `segmentop(segment, address)`? I had it the other way around and the problem went away switching it.

**Comment by astrelsky** [comment]
> Yes the check is done to ignore the unknown SS and allow values to flow in/out of the stack and assumes the segment is the first parameter. WIth out it the complexity of the unknown SS in the segment op calculation would never allow passing the constants in/out of the stack. So they then can be ignored.
> 
> The order of the segment(seg,addr) had been inverted in past years. So it is assumed that the first part will be the segment offset. I had thought that had been corrected quite a while ago. The decompiler might do better analysis on the operations.
> 
> In the code I discovered the issue, a really old win3.1 binary, I don't think it would have worked if the offset would be incorrect and unknown.
> 
> What does your segment operations look like?

This is how it is defined when I encounter an exception.

```xml
<segmentop space="ram" userop="segment" farpointer="yes">
    <pcode>
      <input name="seg" size="1"/>
      <input name="addr" size="2"/>
      <output name="res" size="3"/>
      <body><![CDATA[
        res = (zext(seg) << 16) + zext(addr);
      ]]></body>
    </pcode>
    <constresolve>
      <register name="CES"/>
    </constresolve>
  </segmentop>
```

This is the stack trace for the exception:

```
2022-06-06 08:50:12 ERROR (AnalysisTaskWrapper) Analyzer Error: Analysis Task: Stack - Input parameter size does not match injection specification: pspec: rl78:LE:16:default ghidra.app.plugin.processors.sleigh.SleighException: Input parameter size does not match injection specification: pspec: rl78:LE:16:default
	at ghidra.program.model.lang.InjectPayloadSleigh.checkParameterRestrictions(InjectPayloadSleigh.java:348)
	at ghidra.program.model.lang.InjectPayloadSleigh.setupParameters(InjectPayloadSleigh.java:373)
	at ghidra.program.model.lang.InjectPayloadSleigh.inject(InjectPayloadSleigh.java:168)
	at ghidra.program.model.lang.InjectPayloadSleigh.getPcode(InjectPayloadSleigh.java:188)
	at ghidra.program.util.SymbolicPropogator.doCallOtherPcodeInjection(SymbolicPropogator.java:1668)
	at ghidra.program.util.SymbolicPropogator.applyPcode(SymbolicPropogator.java:967)
	at ghidra.program.util.SymbolicPropogator.flowConstants(SymbolicPropogator.java:504)
	at ghidra.program.util.SymbolicPropogator.flowConstants(SymbolicPropogator.java:400)
	at ghidra.program.util.SymbolicPropogator.flowConstants(SymbolicPropogator.java:193)
	at ghidra.app.cmd.function.NewFunctionStackAnalysisCmd.createStackPointerVariables(NewFunctionStackAnalysisCmd.java:340)
	at ghidra.app.cmd.function.NewFunctionStackAnalysisCmd.analyzeFunction(NewFunctionStackAnalysisCmd.java:191)
	at ghidra.app.cmd.function.NewFunctionStackAnalysisCmd.applyTo(NewFunctionStackAnalysisCmd.java:118)
	at ghidra.app.plugin.core.function.StackVariableAnalyzer.added(StackVariableAnalyzer.java:55)
	at ghidra.app.plugin.core.analysis.AnalysisScheduler.runAnalyzer(AnalysisScheduler.java:186)
	at ghidra.app.plugin.core.analysis.AnalysisTask.applyTo(AnalysisTask.java:39)
	at ghidra.app.plugin.core.analysis.AutoAnalysisManager$AnalysisTaskWrapper.run(AutoAnalysisManager.java:688)
	at ghidra.app.plugin.core.analysis.AutoAnalysisManager.startAnalysis(AutoAnalysisManager.java:788)
	at ghidra.app.plugin.core.analysis.AutoAnalysisManager.startAnalysis(AutoAnalysisManager.java:667)
	at ghidra.app.plugin.core.analysis.AutoAnalysisManager.startAnalysis(AutoAnalysisManager.java:632)
	at ghidra.app.plugin.core.analysis.AnalysisBackgroundCommand.applyTo(AnalysisBackgroundCommand.java:58)
	at ghidra.framework.plugintool.mgr.BackgroundCommandTask.run(BackgroundCommandTask.java:102)
	at ghidra.framework.plugintool.mgr.ToolTaskManager.run(ToolTaskManager.java:336)
	at java.base/java.lang.Thread.run(Thread.java:829)
 ```

**Comment by emteere** [comment]
I see the issue.  Too many assumptions made on the size of varnodes when the whole operation was replaced with a COPY.
The input can't be directly copied into the output unless some size adjustment is factored in.

Would think this would have failed on the orignal test binary, but maybe the sizing just worked.  Good thing I hesitated adding that code in at the last minute when I found the original constant tracking through a segmented stack issue.
